test/common/upstream/eds_subscription_factory_test.cc

```diff
@@ -0,0 +1,108 @@
+using ::testing::Return;
+
+namespace Envoy {
+namespace Upstream {
+
+class EdsSubscriptionFactoryTest : public ::testing::Test {
+public:
+  EdsSubscriptionFactoryTest()
+      :  method_descriptor_(*Protobuf::DescriptorPool::generated_pool()->FindMethodByName(
+          "envoy.api.v2.EndpointDiscoveryService.StreamEndpoints")) {}
+
+  Upstream::MockClusterManager cm_;
+  Event::MockDispatcher dispatcher_;
+  Runtime::MockRandomGenerator random_;
+  Config::MockSubscriptionCallbacks<envoy::api::v2::ClusterLoadAssignment> callbacks_;
+  Stats::MockIsolatedStatsStore stats_store_;
+  NiceMock<LocalInfo::MockLocalInfo> local_info_;
+  Grpc::MockAsyncClient* async_client_;
+  const Protobuf::MethodDescriptor& method_descriptor_;
+  Config::RateLimitSettings rate_limit_settings_;
+};
+
+TEST_F(EdsSubscriptionFactoryTest, ShouldReturnSameMux) {
+  envoy::api::v2::core::ConfigSource config;
+  config.mutable_api_config_source()->add_grpc_services()->mutable_envoy_grpc()->set_cluster_name("first_cluster");
+  EdsSubscriptionFactory factory;
+
+  EXPECT_CALL(dispatcher_, createTimer_(_));
+  Config::GrpcMux& first_mux = factory.getOrCreateMux(local_info_, std::unique_ptr<Grpc::MockAsyncClient>(new Grpc::MockAsyncClient()), dispatcher_, method_descriptor_,
+                                    random_, config.api_config_source(), stats_store_, rate_limit_settings_);
+  Config::GrpcMux& second_mux = factory.getOrCreateMux(local_info_, std::unique_ptr<Grpc::MockAsyncClient>(new Grpc::MockAsyncClient()), dispatcher_, method_descriptor_,
+                                                    random_, config.api_config_source(), stats_store_, rate_limit_settings_);
+  ASSERT_TRUE(&first_mux == &second_mux);
+}
+
+TEST_F(EdsSubscriptionFactoryTest, ShouldReturnDifferentMuxes) {
+  EdsSubscriptionFactory factory;
+
+  EXPECT_CALL(dispatcher_, createTimer_(_)).Times(2);
+  envoy::api::v2::core::ConfigSource first_config;
+  first_config.mutable_api_config_source()->add_grpc_services()->mutable_envoy_grpc()->set_cluster_name("first_cluster");
+  Config::GrpcMux& first_mux = factory.getOrCreateMux(local_info_, std::unique_ptr<Grpc::MockAsyncClient>(new Grpc::MockAsyncClient()), dispatcher_, method_descriptor_,
+                                                      random_, first_config.api_config_source(), stats_store_, rate_limit_settings_);
+  envoy::api::v2::core::ConfigSource second_config;
+  second_config.mutable_api_config_source()->add_grpc_services()->mutable_envoy_grpc()->set_cluster_name("second_cluster");
+  Config::GrpcMux& second_mux = factory.getOrCreateMux(local_info_, std::unique_ptr<Grpc::MockAsyncClient>(new Grpc::MockAsyncClient()), dispatcher_, method_descriptor_,
+                                                       random_, second_config.api_config_source(), stats_store_, rate_limit_settings_);
+  ASSERT_TRUE(&first_mux != &second_mux);
+}
+
+TEST_F(EdsSubscriptionFactoryTest, ShouldUseGetOrCreateMuxWhenApiConfigSourceIsUsed) {
```

Comment:
Can you add a single line `//` above each test explaining what it does in natural language?

Reply:
done.